lib/path/default.nix

```diff
@@ -190,6 +193,80 @@ in /* No rec! Add dependencies on this file at the top. */ {
     subpathInvalidReason value == null;
 
 
+  /* Join subpath strings together using `/`, returning a normalised subpath string.
+
+    Like `concatStringsSep "/"` but safer, specifically:
+
+    - All elements must be valid subpath strings, see `lib.path.subpath.isValid`
+
+    - The result gets normalised, see `lib.path.subpath.normalise`
+
+    - The edge case of an empty list gets properly handled by returning the neutral subpath `"./."`
+
+    Laws:
+
+    - Associativity:
+
+          subpath.join [ x (subpath.join [ y z ]) ] == subpath.join [ (subpath.join [ x y ]) z ]
+
+    - Identity - `"./."` is the neutral element for normalised paths:
+
+          subpath.join [ ] == "./."
+          subpath.join [ (subpath.normalise p) "./." ] == subpath.normalise p
+          subpath.join [ "./." (subpath.normalise p) ] == subpath.normalise p
+
+    - Normalisation - the result is normalised according to `lib.path.subpath.normalise`:
+
+          subpath.join ps == subpath.normalise (subpath.join ps)
+
+    - For non-empty lists, the implementation is equivalent to normalising the result of `concatStringsSep "/"`.
+      Note that the above laws can be derived from this one.
+
+          ps != [] -> subpath.join ps == subpath.normalise (concatStringsSep "/" ps)
+
+    Type:
+      subpath.join :: [ String ] -> String
+
+    Example:
+      subpath.join [ "foo" "bar/baz" ]
+      => "./foo/bar/baz"
+
+      # normalise the result
+      subpath.join [ "./foo" "." "bar//./baz/" ]
+      => "./foo/bar/baz"
+
+      # passing an empty list results in the current directory
+      subpath.join [ ]
+      => "./."
+
+      # elements must be valid subpath strings
+      subpath.join [ /foo ]
+      => <error>
+      subpath.join [ "" ]
+      => <error>
+      subpath.join [ "/foo" ]
+      => <error>
+      subpath.join [ "../foo" ]
+      => <error>
+  */
+  subpath.join =
+    # The list of subpaths to join together
+    subpaths:
+    assert assertMsg
+      # Fast in case all paths are valid
+      (all isValid subpaths)
+      # Otherwise we take our time to gather more info for a better error message
+      # Strictly go through each path, throwing on the first invalid one
+      # Tracks the list index in the fold accumulator
+      (foldl' (i: path:
+        if isValid path
+        then i + 1
+        else throw ''
+          lib.path.subpath.join: Element at index ${toString i} is not a valid subpath string:
+              ${subpathInvalidReason path}''
+      ) 0 subpaths);
```

Comment:
I also simplified this a bit to remove the `seq` and `abort` fallthru thing, I think that was overengineered.

Reply:
I really don't like the abuse of the exception "side effect" here.
It starts with `assert`, which is never triggered and then it continues with `assertMsg`, which is also never triggered. The final error happens to be in the error message of `assertMsg`, evaluated when it almost triggers.
This is enough reason to just use an `if` expression instead.

Final nitpick: assertions are for checking the invariants of self-contained software units, and not for input validation across a public interface. See [assert (verb)](https://en.wiktionary.org/wiki/assert)
> 1\. To [declare](https://en.wiktionary.org/wiki/declare) with [assurance](https://en.wiktionary.org/wiki/assurance) or plainly and strongly; to state positively.
> [...]
> (programming) To specify that a condition or expression is true at a certain point in the code.

Reply:
Good point, using an `if` now.

Regarding assertions in general, yeah maybe `assert` is being abused in that way in Nix, just because it's somewhat easier to type. For now it seems to be the pattern though, so I still think that with https://github.com/NixOS/nixpkgs/pull/215166 it's okay to replace this `if` here with an `assert assertAll isValid subpaths`